✨ RUM Browser Vue Integration (#2962)
DataDog/integrations-extras: rum_vue/README.md

# RUM Browser Monitoring - Vue integration

## Overview

The Datadog RUM Vue integration provides framework-specific instrumentation to help you monitor and debug Vue 3 applications. This integration adds:

- **Automatic route change detection** using Vue Router v4

> - **View name normalization** that maps dynamic route segments to their parameterized definitions (e.g. `/users/123` becomes `/users/:id`)

Reviewer: ```suggestion
- **View name normalization** that maps dynamic route segments to their parameterized definitions (for example, `/users/123` becomes `/users/:id`)
```

- **Error reporting** through Vue's global error handler with full component stack traces
- **Full-stack visibility** by correlating frontend performance with backend traces and logs

Combined with Datadog RUM's core capabilities, you can debug performance bottlenecks, track user journeys, monitor Core Web Vitals, and analyze every user session with context.

## Setup

> Start by setting up [Datadog RUM][1] in your Vue application:

Reviewer: ```suggestion
Set up [Datadog RUM][1] in your Vue application:
```

> - If you are creating a RUM application, select **Vue** as the application type.

Reviewer: Clarifying question: if they were to select Vue as the application type, do they need to also follow the steps below? Do they only follow the steps below if they select JavaScript?

Author: If you select Vue, the Datadog App guides you through the setup. The manual steps below are only needed if Vue is not available and you select JavaScript instead. I updated the text to make this clearer.

- If Vue is not available as an option, select **JavaScript** and follow the steps below to integrate the plugin manually.

> After configuration, the Datadog App provides instructions for integrating the [RUM-Vue plugin][2] with the Browser SDK.

Reviewer: ```suggestion
After you configure your application, see Datadog for instructions on integrating the [RUM-Vue plugin][2] with the Browser SDK.
```

> This integration requires **Vue v3.5+** and **Vue Router v4+** (if using router view tracking).

Reviewer: ```suggestion
This integration requires **Vue v3.5+**, or **Vue Router v4+** if you are using router view tracking.
```
Does this keep the meaning? I understand this as the default is Vue v3.5+, but you need Vue Router v4+ if you are using router view tracking.

Author: Vue v3.5+ is always required. Vue Router v4+ is only an additional requirement when using router tracking. Changing to "or" makes it sound like they are alternatives. I reverted to the original wording.

## Basic usage

> 

Reviewer: For your consideration
I'd suggest adding a short introductory sentence(s) here before the level 3 headers -- similar to the  ##Router view tracking section

Author: Added!

### 1. Initialize the Datadog RUM SDK with the Vue plugin

> In your `main.ts` (or `main.js`):

Reviewer: ```suggestion
Add the following to your `main.ts` (or `main.js`) file:
```

```javascript
import { datadogRum } from '@datadog/browser-rum'
import { vuePlugin } from '@datadog/browser-rum-vue'

datadogRum.init({
  applicationId: '<APP_ID>',
  clientToken: '<CLIENT_TOKEN>',
  site: 'datadoghq.com',
  plugins: [vuePlugin()],
})
```

### 2. Attach the Vue error handler

Use `addVueError` as your application's global error handler to automatically report Vue errors to Datadog RUM with component stack traces:

```javascript
import { createApp } from 'vue'
import { addVueError } from '@datadog/browser-rum-vue'
import App from './App.vue'

const app = createApp(App)
app.config.errorHandler = addVueError
app.mount('#app')
```

## Router view tracking

> To automatically track route changes as RUM views, enable the `router` option in the plugin and use the Datadog `createRouter` wrapper instead of Vue Router's native one.

Reviewer: ```suggestion
To automatically track route changes as RUM views, enable the `router` option in the plugin then replace the Vue Router's native wrapper with Datadog's `createRouter` wrapper.
```
Does this keep the meaning? Adding some suggestions for clarification.

### 1. Initialize the RUM SDK with router tracking enabled

```javascript
import { datadogRum } from '@datadog/browser-rum'
import { vuePlugin } from '@datadog/browser-rum-vue'

datadogRum.init({
  applicationId: '<APP_ID>',
  clientToken: '<CLIENT_TOKEN>',
  site: 'datadoghq.com',
  plugins: [vuePlugin({ router: true })],
})
```

### 2. Create your router with the Datadog wrapper

> Replace Vue Router's `createRouter` with the one from `@datadog/browser-rum-vue/vue-router-v4`:

Reviewer: ```suggestion
Import the `createRouter` from `@datadog/browser-rum-vue/vue-router-v4` instead of from `vue-router`:
```
Does this keep the meaning or does it rid of any nuance?

```javascript
import { createWebHistory } from 'vue-router'
import { createRouter } from '@datadog/browser-rum-vue/vue-router-v4'

const router = createRouter({
  history: createWebHistory(),
  routes: [
    { path: '/', component: Home },
    { path: '/users', component: Users },
    { path: '/users/:id', component: UserDetail },
  ],
})
```

### 3. Wire everything together

```javascript
import { createApp } from 'vue'
import { createWebHistory } from 'vue-router'
import { datadogRum } from '@datadog/browser-rum'
import { vuePlugin, addVueError } from '@datadog/browser-rum-vue'
import { createRouter } from '@datadog/browser-rum-vue/vue-router-v4'
import App from './App.vue'

datadogRum.init({
  applicationId: '<APP_ID>',
  clientToken: '<CLIENT_TOKEN>',
  site: 'datadoghq.com',
  plugins: [vuePlugin({ router: true })],
})

const router = createRouter({
  history: createWebHistory(),
  routes: [
    { path: '/', component: Home },
    { path: '/users', component: Users },
    { path: '/users/:id', component: UserDetail },
  ],
})

const app = createApp(App)
app.config.errorHandler = addVueError
app.use(router)
app.mount('#app')
```

> ## Route tracking

Reviewer: Should this section go to the section under line 55?
It seems ## Route tracking and ## Router view tracking might be discussing similar things? Would it make sense to add this section to the one above?

Let me know if that makes sense or if I may have misunderstood a nuance here.

Author: They cover different things. "Router view tracking" is the setup (how to enable it), "Route tracking" is a reference for URL normalization behavior (what it does once enabled). The Next.js README follows the same separation. What do you think?
